solutions/USACO open20-haircut.mdx

```diff
@@ -0,0 +1,165 @@
+---
+id: usaco-1041
+source: USACO Gold 2020 US Open
+title: Haircut
+author: Jeffrey Meng
+---
+
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_haircut_gold_open20.html)
+
+First, we need to count the number of inversions. (Fun fact: the number of inversions is also a measure of how sorted an array is. If an array has 0 inversions, it's perfectly sorted in ascending order. Conversely, if it has the maximum possible number of inversions, the array is perfectly sorted in descending order.) 
+
+One way to count inversions is to use a binary indexed tree. Think of an inversion as a pair of values $a, b$, with $a > b$ and $a$ appearring before $b$ in the array. Then, we can easily compute, for each value in the array, the number of inversions for which it is $b$.
+
+To do this, we use a BIT as a frequency table, keeping track of the number of hairs of each possible value. If our BIT is named `tree`, `tree[i]` will tell us how many times $i$ is seen in the array of hairs. The trick is that as long as we fill out the frequency table in the order the elements appear in the array, we can use the frequency table to also count the number of inversions, because at that point in time the frequency table only contains hairs to the left of the hair we are currently considering.
```

Comment:
```suggestion
To do this, we use a BIT as a frequency table, keeping track of the number of hairs of each possible value. If our BIT is named `tree`, `tree[i]` will tell us how many times $i$ is seen in the array of hairs. The trick is that as long as we fill out the frequency table in the order the elements appear in the array, we can use the frequency table also to count the number of inversions because the frequency table only contains hairs to the left of the hair we are currently considering at that point in time.
```